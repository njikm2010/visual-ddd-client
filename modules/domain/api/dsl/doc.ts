import { Doc as YDoc, XmlElement, XmlText, Map as YMap, Array as YArray } from 'yjs';
import { buildEditorYjs, buildEmptyEditorYjs } from '@/lib/editor/Model/Yjs';
import { YjsDocMetaInfo } from '@/lib/yjs-store-api';
import { AggregationDSL, EntityDSL, TypeType, CommandDSL } from '@/modules/domain/domain-design/dsl/dsl';
import { buildUbiquitousLanguageYjs } from '@/modules/domain/ubiquitous-language-design/Yjs';
import { toNameCase } from '@/lib/utils';

import { YJS_FIELD_NAME } from '../../constants';

import * as DSL from './interface';
import { DSLModel } from './model';

/**
 * 创建业务域模型
 */
function createDomainModel(meta: YjsDocMetaInfo, map: YMap<unknown>) {
  const { name, title, description } = meta;

  buildEditorYjs(
    {
      nodes: [
        // 业务活动
        {
          id: '{activity-create}',
          name: 'activity',
          properties: {
            label: `新增${title}`,
            position: { x: 171, y: 181 },
            size: { width: 120, height: 75 },
          },
        },
        {
          id: '{activity-update}',
          name: 'activity',
          properties: {
            label: `编辑${title}`,
            position: { x: 443, y: 183.5 },
            size: { width: 120, height: 75 },
          },
        },
        {
          id: '{activity-delete}',
          name: 'activity',
          properties: {
            label: `删除${title}`,
            size: { width: 120, height: 70 },
            position: { x: 743, y: 183.5 },
          },
        },
        // 聚合
        {
          id: '{aggregation}',
          name: 'aggregation',
          properties: helper => {
            return {
              zIndex: 1,
              position: { x: 171, y: 715 },
              size: { width: 500, height: 500 },
              __prevent_auto_resize__: true,
              ...({
                uuid: helper.getId('aggregation'),
                name: toNameCase('CamelCase', name),
                title: title,
                description: description,
                meta: [],
                color: '#D9F7BE',
              } satisfies AggregationDSL),
            };
          },
          children: [
            {
              id: '{entity}',
              name: 'entity',
              properties: helper => ({
                position: { x: 231, y: 775 },
                zIndex: 2,
                size: { width: 200, height: 144.828125 },
                ...({
                  uuid: helper.getId('entity'),
                  name: toNameCase('CamelCase', name),
                  title: title,
                  description: description,
                  meta: [],
                  implements: [],
                  abstract: false,
                  methods: [],
                  classProperties: [],
                  classMethods: [],
                  isAggregationRoot: true,
                  id: helper.getOrCreateId('entity-id'),
                  properties: [
                    {
                      uuid: helper.getId('entity-id'),
                      name: 'id',
                      title: `${title} ID`,
                      description: '',
                      meta: [],
                      type: { type: TypeType.Base, name: 'Long' },
                      access: 'public',
                    },
                    {
                      uuid: helper.getOrCreateId(),
                      name: 'name',
                      title: '名称',
                      description: '',
                      meta: [],
                      type: { type: TypeType.Base, name: 'String' },
                      access: 'public',
                    },
                  ],
                } satisfies EntityDSL),
              }),
            },
          ],
        },

        // 命令
        {
          id: '{command-create}',
          name: 'command',
          properties: helper => ({
            position: { x: 106, y: 350 },
            zIndex: 2,
            size: { width: 200, height: 184.671875 },
            ...({
              uuid: helper.getId('command-create'),
              name: toNameCase('CamelCase', name) + 'Create',
              title: `新增${title}`,
              description: '',
              meta: [],
              repository: 'create',
              eventSendable: false,
              aggregation: { referenceId: helper.getId('aggregation') },
              category: 'create',
              source: {
                http: { enabled: true },
                rpc: { enabled: true },
                event: { enabled: false, value: '' },
                schedule: { enabled: false, value: '' },
              },
              properties: [
                {
                  uuid: helper.getOrCreateId(),
                  name: 'name',
                  title: '名称',
                  description: '',
                  meta: [],
                  type: { type: TypeType.Base, name: 'String' },
                  access: 'public',
                },
              ],
              eventProperties: [
                {
                  uuid: helper.getOrCreateId(),
                  name: 'id',
                  title: `${title} ID`,
                  description: '',
                  meta: [],
                  type: { type: TypeType.Base, name: 'Long' },
                  access: 'public',
                },
              ],
            } satisfies CommandDSL),
          }),
        },
        {
          id: '{command-modify}',
          name: 'command',
          properties: helper => ({
            position: { x: 425, y: 350 },
            size: { width: 200, height: 183.671875 },
            zIndex: 2,
            ...({
              uuid: helper.getId('command-modify'),
              name: toNameCase('CamelCase', name) + 'Modify',
              title: `编辑${title}`,
              description: '',
              meta: [],
              repository: 'modify',
              eventSendable: false,
              aggregation: { referenceId: helper.getId('aggregation') },
              category: 'modify',
              source: {
                http: { enabled: true },
                rpc: { enabled: true },
                event: { enabled: false, value: '' },
                schedule: { enabled: false, value: '' },
              },
              properties: [
                {
                  uuid: helper.getOrCreateId(),
                  name: 'id',
                  title: `${title} ID`,
                  description: '',
                  meta: [],
                  type: { type: TypeType.Base, name: 'Long' },
                  access: 'public',
                },
                {
                  uuid: helper.getOrCreateId(),
                  name: 'name',
                  title: '名称',
                  description: '',
                  meta: [],
                  type: { type: TypeType.Base, name: 'String' },
                  access: 'public',
                },
              ],
              eventProperties: [],
            } satisfies CommandDSL),
          }),
        },
        {
          id: '{command-remove}',
          name: 'command',
          properties: helper => ({
            position: { x: 743, y: 350 },
            size: { width: 200, height: 129.96875 },
            zIndex: 2,
            ...({
              uuid: helper.getId('command-remove'),
              name: toNameCase('CamelCase', name) + 'Remove',
              title: `删除${title}`,
              description: '',
              meta: [],
              repository: 'remove',
              eventSendable: false,
              aggregation: { referenceId: helper.getId('aggregation') },
              category: 'remove',
              source: {
                http: { enabled: true },
                rpc: { enabled: true },
                event: { enabled: false, value: '' },
                schedule: { enabled: false, value: '' },
              },
              properties: [
                {
                  uuid: helper.getOrCreateId(),
                  name: 'id',
                  title: `${title} ID`,
                  description: '',
                  meta: [],
                  type: { type: TypeType.Base, name: 'Long' },
                  access: 'public',
                },
              ],
              eventProperties: [],
            } satisfies CommandDSL),
          }),
        },
      ],
      edges: [
        {
          name: 'edge',
          source: {
            cell: '{activity-create}',
            port: 'right',
          },
          target: {
            cell: '{activity-update}',
            port: 'left',
          },
        },
        {
          name: 'edge',
          source: {
            cell: '{activity-update}',
            port: 'right',
          },
          target: {
            cell: '{activity-delete}',
            port: 'left',
          },
        },
      ],
    },
    map
  );
}

function createUbiquitousLanguages(array: YArray<YMap<string>>) {
  buildUbiquitousLanguageYjs(
    [
      { conception: '资源', englishName: 'Resource', definition: 'BD进行销售活动的对象', example: '', restraint: '' },
      {
        conception: '公海',
        englishName: 'PublicSea',
        definition: '业务目标限定的模拟资源全集',
        example: '业务目标是团购买单合作，限定POI为到餐全国和综合58同城的',
        restraint: '与业务目标对应，比如交易公海是为了做交易，推广公海是为了推广',
      },
    ],

    array
  );
}

export function createDoc(params: YjsDocMetaInfo) {
  const doc = new YDoc();

  const domainMap = doc.getMap(YJS_FIELD_NAME.DOMAIN);

  createDomainModel(params, domainMap);

  // 图形编辑器
  // 初始化根节点
  [YJS_FIELD_NAME.QUERY, YJS_FIELD_NAME.DATA_OBJECT, YJS_FIELD_NAME.MAPPER].forEach(name => {
    const map = doc.getMap(name);
    buildEmptyEditorYjs(map);
  });

  // 产品文档
  const product = doc.getXmlFragment(YJS_FIELD_NAME.PRODUCT);
  const productTemplate = new XmlElement('heading');
  productTemplate.setAttribute('level', '1');
  productTemplate.insert(0, [new XmlText(`${params.title}产品文档`)]);
  product.insert(0, [productTemplate]);

  // 产品愿景
  doc.getText(YJS_FIELD_NAME.VISION);

  // 统一语言
  const arr = doc.getArray<YMap<string>>(YJS_FIELD_NAME.UBIQUITOUS_LANGUAGE);
  createUbiquitousLanguages(arr);

  return doc;
}

/**
 * 转换为 DSL
 * @param doc
 * @returns
 */
export function transformToDSL(doc: YDoc): DSL.BusinessDomainDSL {
  const model = new DSLModel(doc);

  return model.toDSL();
}
