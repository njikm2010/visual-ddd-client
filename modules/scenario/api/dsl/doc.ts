import { Doc as YDoc } from 'yjs';
import { buildEditorYjs, buildEmptyEditorYjs } from '@/lib/editor/Model/Yjs';
import { NoopArray } from '@wakeapp/utils';

import { ScenarioObjectName } from '../../scenario-design/dsl/constants';
import type { LanesDSL } from '../../scenario-design/dsl/dsl';
import { YJS_FIELD_NAME } from '../../constants';

import { DSLModel } from './model';
import { ScenarioDSL } from './interface';

/**
 * 创建初始化模板
 * @returns
 */
export function createDoc() {
  const doc = new YDoc();

  // 初始化业务场景泳道图
  const scenarioMap = doc.getMap(YJS_FIELD_NAME.SCENARIO);

  buildEditorYjs(
    {
      nodes: [
        {
          id: '{lane}',
          name: ScenarioObjectName.Lane,
          properties: helper => {
            return {
              position: { x: 10, y: 10 },
              ...({
                uuid: helper.getId('lane'),
                name: 'lane',
                title: '泳道',
                description: '',
                meta: [],
                width: 1200,
                panes: [{ uuid: helper.getOrCreateId(), title: '未命名泳道', height: 300 }],
              } satisfies LanesDSL),
            };
          },
        },
      ],
      edges: NoopArray,
    },
    scenarioMap
  );

  // 图形编辑器
  // 初始化根节点
  [YJS_FIELD_NAME.SERVICE].forEach(name => {
    const map = doc.getMap(name);
    buildEmptyEditorYjs(map);
  });

  return doc;
}

/**
 * 转换为 DSL
 * @param doc
 * @returns
 */
export function transformToDSL(doc: YDoc): ScenarioDSL {
  const model = new DSLModel(doc);

  return model.toDSL();
}
