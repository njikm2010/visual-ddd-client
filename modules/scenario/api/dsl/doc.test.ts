import { createDoc } from './doc';
import * as mockedUuid from 'uuid';

jest.mock('uuid', () => {
  let count = 0;
  return {
    __esModule: true,
    reset() {
      count = 0;
    },
    v4() {
      return `uuid-${count++}`;
    },
  };
});

const mockedUuidModule = mockedUuid as unknown as jest.Mocked<{ reset(): void }>;

beforeEach(() => {
  mockedUuidModule.reset();
});

test('createDoc', () => {
  const doc = createDoc();

  expect(doc.toJSON()).toMatchSnapshot();
});
