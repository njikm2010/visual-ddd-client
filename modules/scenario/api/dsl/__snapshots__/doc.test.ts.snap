// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDoc 1`] = `
{
  "scenario": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {
        "uuid-0": 1,
      },
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
    "uuid-0": {
      "__NODE__": true,
      "children": {},
      "id": "uuid-0",
      "locked": undefined,
      "parent": "__ROOT__",
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "s-lane",
        "__node_type__": "node",
        "description": "",
        "meta": [],
        "name": "lane",
        "panes": [
          {
            "height": 300,
            "title": "未命名泳道",
            "uuid": "uuid-1",
          },
        ],
        "position": {
          "x": 10,
          "y": 10,
        },
        "title": "泳道",
        "uuid": "uuid-0",
        "width": 1200,
      },
    },
  },
  "service": {
    "__ROOT__": {
      "__NODE__": true,
      "children": {},
      "id": "__ROOT__",
      "locked": undefined,
      "parent": undefined,
      "properties": {
        "__PROPERTY__": true,
        "__node_name__": "__ROOT__",
        "__node_type__": "node",
      },
    },
  },
}
`;
